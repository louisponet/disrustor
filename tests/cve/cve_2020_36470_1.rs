#![forbid(unsafe_code)]

use disrustor::internal::RingBuffer;

use std::marker::PhantomData;
use std::thread;

struct NonSend {
    created_thread: thread::ThreadId,
    // Mark this struct as `NonSend`
    _marker: PhantomData<*mut ()>,
}

impl Default for NonSend {
    fn default() -> Self {
        NonSend {
            created_thread: thread::current().id(),
            _marker: PhantomData,
        }
    }
}

impl Drop for NonSend {
    fn drop(&mut self) {
        if thread::current().id() != self.created_thread {
            panic!("NonSend destructor is running on a wrong thread!");
        }
    }
}

fn main() {
    let buffer = RingBuffer::<NonSend>::new(1);

    let handle = thread::spawn(move || {
        drop(buffer);
    });

    handle.join().unwrap();
}