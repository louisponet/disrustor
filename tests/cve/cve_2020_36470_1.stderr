error[E0277]: `*mut ()` cannot be sent between threads safely
  --> tests/cve/cve_2020_36470_1.rs:34:32
   |
34 |       let handle = thread::spawn(move || {
   |  __________________-------------_^
   | |                  |
   | |                  required by a bound introduced by this call
35 | |         drop(buffer);
36 | |     });
   | |_____^ `*mut ()` cannot be sent between threads safely
   |
   = help: within `NonSend`, the trait `Send` is not implemented for `*mut ()`
   = note: required because it appears within the type `PhantomData<*mut ()>`
note: required because it appears within the type `NonSend`
  --> tests/cve/cve_2020_36470_1.rs:8:8
   |
8  | struct NonSend {
   |        ^^^^^^^
   = note: required for `RingBuffer<NonSend>` to implement `Send`
note: required because it's used within this closure
  --> tests/cve/cve_2020_36470_1.rs:34:32
   |
34 |     let handle = thread::spawn(move || {
   |                                ^^^^^^^
note: required by a bound in `std::thread::spawn`
  --> $RUST/std/src/thread/mod.rs
   |
   |     F: Send + 'static,
   |        ^^^^ required by this bound in `spawn`
